--- perl-5.10.1.orig/perl.c	2009-04-22 20:07:41.000000000 +0200
+++ perl-5.10.1/perl.c	2010-11-26 22:06:52.225163765 +0100
@@ -89,20 +89,22 @@
 #ifdef NETWARE
 #include "nwutil.h"	
 char *nw_get_sitelib(const char *pl);
 #endif
 
 /* XXX If this causes problems, set i_unistd=undef in the hint file.  */
 #ifdef I_UNISTD
 #include <unistd.h>
 #endif
 
+#include "mini_preamble.h"
+
 #ifdef DEBUG_LEAKING_SCALARS_FORK_DUMP
 #  ifdef I_SYS_WAIT
 #   include <sys/wait.h>
 #  endif
 #  ifdef I_SYSUIO
 #    include <sys/uio.h>
 #  endif
 
 union control_un {
   struct cmsghdr cm;
@@ -1511,20 +1513,23 @@
 #else
 #ifdef OS2
 	sv_setpv(GvSVn(tmpgv), os2_execname(aTHX));
 #else
 	sv_setpv(GvSVn(tmpgv),PL_origargv[0]);
 #endif
 #endif
     }
 }
 
+char do_add_mini_preamble = 0;
+
+
 /*
 =for apidoc perl_parse
 
 Tells a Perl interpreter to parse a Perl script.  See L<perlembed>.
 
 =cut
 */
 
 int
 perl_parse(pTHXx_ XSINIT_t xsinit, int argc, char **argv, char **env)
@@ -1556,20 +1561,25 @@
 	const char * const s = PerlEnv_getenv("PERL_HASH_SEED_DEBUG");
 
 	if (s && (atoi(s) == 1))
 	    PerlIO_printf(Perl_debug_log, "HASH_SEED = %"UVuf"\n", PL_rehash_seed);
     }
 #endif /* #if defined(USE_HASH_SEED) || defined(USE_HASH_SEED_EXPLICIT) */
 
     PL_origargc = argc;
     PL_origargv = argv;
 
+    if (do_add_mini_preamble) {
+        /* SUXX: an unquoted newline is not allowed here (in PL_Preambleav) */
+        Perl_av_create_and_push(aTHX_ &PL_preambleav, newSVpvn(mini_preamble, sizeof mini_preamble));
+    }
+
     if (PL_origalen != 0) {
 	PL_origalen = 1; /* don't use old PL_origalen if perl_parse() is called again */
     }
     else {
 	/* Set PL_origalen be the sum of the contiguous argv[]
 	 * elements plus the size of the env in case that it is
 	 * contiguous with the argv[].  This is used in mg.c:Perl_magic_set()
 	 * as the maximum modifiable length of $0.  In the worst case
 	 * the area we are able to modify is limited to the size of
 	 * the original argv[0].  (See below for 'contiguous', though.)
--- perl-5.10.1.orig/miniperlmain.c	2009-02-12 23:58:17.000000000 +0100
+++ perl-5.10.1/miniperlmain.c	2010-11-26 22:27:49.474851364 +0100
@@ -41,20 +41,22 @@
 
 static void xs_init (pTHX);
 static PerlInterpreter *my_perl;
 
 #if defined (__MINT__) || defined (atarist)
 /* The Atari operating system doesn't have a dynamic stack.  The
    stack size is determined from this value.  */
 long _stksize = 64 * 1024;
 #endif
 
+extern char do_add_mini_preamble;
+
 #if defined(PERL_GLOBAL_STRUCT_PRIVATE)
 /* The static struct perl_vars* may seem counterproductive since the
  * whole idea PERL_GLOBAL_STRUCT_PRIVATE was to avoid statics, but note
  * that this static is not in the shared perl library, the globals PL_Vars
  * and PL_VarsPtr will stay away. */
 static struct perl_vars* my_plvarsp;
 struct perl_vars* Perl_GetVarsPrivate(void) { return my_plvarsp; }
 #endif
 
 #ifdef NO_ENV_ARRAY_IN_MAIN
@@ -68,20 +70,21 @@
 {
     dVAR;
     int exitstatus;
 #ifdef PERL_GLOBAL_STRUCT
     struct perl_vars *plvarsp = init_global_struct();
 #  ifdef PERL_GLOBAL_STRUCT_PRIVATE
     my_vars = my_plvarsp = plvarsp;
 #  endif
 #endif /* PERL_GLOBAL_STRUCT */
     (void)env;
+    do_add_mini_preamble = 0;
 #ifndef PERL_USE_SAFE_PUTENV
     PL_use_safe_putenv = 0;
 #endif /* PERL_USE_SAFE_PUTENV */
 
     /* if user wants control of gprof profiling off by default */
     /* noop unless Configure is given -Accflags=-DPERL_GPROF_CONTROL */
     PERL_GPROF_MONCONTROL(0);
 
 #ifdef NO_ENV_ARRAY_IN_MAIN
     PERL_SYS_INIT3(&argc,&argv,&environ);
--- perl-5.10.1.orig/Configure	2009-08-18 21:03:53.000000000 +0200
+++ perl-5.10.1/Configure	2010-11-26 21:00:01.284891584 +0100
@@ -5139,25 +5139,25 @@
 		eval $checkccflag
 		;;
 	esac
 
 	# on x86_64 (at least) we require an extra library (libssp) in the
 	# link command line. This library is not named, so I infer that it is
 	# an implementation detail that may change. Hence the safest approach
 	# is to add the flag to the flags passed to the compiler at link time,
 	# as that way the compiler can do the right implementation dependant
 	# thing. (NWC)
-	case "$gccversion" in
-	?*)	set stack-protector -fstack-protector
-		eval $checkccflag
-		;;
-	esac
+	#case "$gccversion" in
+	#?*)	set stack-protector -fstack-protector
+	#	eval $checkccflag
+	#	;;
+	#esac
 	;;
 esac
 
 case "$mips_type" in
 *BSD*|'') inclwanted="$locincpth $usrinc";;
 *) inclwanted="$locincpth $inclwanted $usrinc/bsd";;
 esac
 for thisincl in $inclwanted; do
 	if $test -d $thisincl; then
 		if $test x$thisincl != x$usrinc; then
@@ -11987,22 +11987,23 @@
 : see if prototype for drand48 is available
 echo " "
 set d_drand48proto drand48 $i_stdlib stdlib.h $i_unistd unistd.h
 eval $hasproto
 
 : see if dup2 exists
 set dup2 d_dup2
 eval $inlibc
 
 : see if eaccess exists
-set eaccess d_eaccess
-eval $inlibc
+#set eaccess d_eaccess
+#eval $inlibc
+d_eaccess=undef
 
 : see if endgrent exists
 set endgrent d_endgrent
 eval $inlibc
 
 : see if this is an grp system
 set grp.h i_grp
 eval $inhdr
 
 case "$i_grp" in
